.circleci/config.yml

```diff
@@ -1,34 +1,114 @@
     docker:
-    - image: cypress/base:10
-      environment:
-          ## this enables colors in the output
-        TERM: xterm
-      auth:
-        username: $DOCKER_HUB_USERNAME
-        password: $DOCKER_HUB_PASSWORD
+      - image: cypress/base:10
+        environment:
+          TERM: xterm
+        auth:
+          username: $DOCKER_HUB_USERNAME
+          password: $DOCKER_HUB_PASSWORD
+
+
+workflows:
+  version: 2
+  build:
+    jobs:
+      - setup
+      - quality-gate-build:
+          requires:
+            - setup
+      - quality-gate-lint:
+          requires:
+            - setup
+      - quality-gate-test:
+          requires:
+            - setup
+      - quality-gate-e2e:
+          requires:
+            - setup
+
+jobs:
+  setup:
+    executor: node
+    steps:
+      - checkout
+      - *restore_yarn_cache
+      - *restore_node_modules
+      - run:
+          name: Installing dependencies
+          command: yarn --frozen-lockfile --cache-folder ~/.cache/yarn
+      - save_cache:
+          # Store node_modules for all jobs in this workflow so that they don't
+          # need to each run a yarn install for each job. This will speed up
+          # all jobs run on this branch with the same lockfile.
+          name: Save node_modules cache
+          key: v1-node-modules-{{ arch }}-{{ .Branch }}-{{ checksum "yarn.lock" }}
+          paths:
+            - node_modules
+      - persist_to_workspace:
+          root: .
+          paths:
+            - node_modules
+  quality-gate-build:
+    executor: node
+    steps:
+      - checkout
+      - attach_workspace:
+          at: .
+      - run:
+          name: Run build
+          command: yarn build
+  quality-gate-lint:
+    executor: node
+    steps:
+      - checkout
+      - attach_workspace:
+          at: .
+      - run:
+          name: Run linting
+          command: yarn lint
+  quality-gate-test:
+    executor: node
+    steps:
+      - checkout
+      - attach_workspace:
+          at: .
+      - run:
+          name: Run unit tests
+          command: yarn unit
+  quality-gate-e2e:
+    executor: cypress
     steps:
-    - checkout
-    - restore_cache:   # special step to restore the dependency cache
-        key: dependency-cache-{{ checksum "yarn.lock" }}
-    - run:
-        name: Setup Dependencies
-        command: yarn install --frozen-lockfile
-    - save_cache:   # special step to save the dependency cache
-        key: dependency-cache-{{ checksum "yarn.lock" }}
-        paths:
-        - ~/.cache     ## cache both yarn and Cypress
-    - run:
-        name: Run linting
-        command: |
-          yarn lint
-    - run:
-        name: Run unit tests
-        command: |
-          yarn unit
-    - run:   # run coveralls
-        name: Run e2e tests
-        command: |
-          yarn e2e:ci
+      - checkout
+      - attach_workspace:
+          at: .
+      - run:
+          name: Installing Cypress Binary
+          command: yarn cypress install
```

Comment:
This was needed because of the following error: 

```
The cypress npm package is installed, but the Cypress binary is missing.

We expected the binary to be installed here: /root/.cache/Cypress/5.2.0/Cypress/Cypress

Reasons it may be missing:

- You're caching 'node_modules' but are not caching this path: /root/.cache/Cypress
- You ran 'npm install' at an earlier build step but did not persist: /root/.cache/Cypress

Properly caching the binary will fix this error and avoid downloading and unzipping Cypress.

Alternatively, you can run 'cypress install' to download the binary again.

https://on.cypress.io/not-installed-ci-error
```

If you have time, we can also try caching this during `setup` and avoid double installing:  
![image](https://user-images.githubusercontent.com/1442690/143878781-c23409a9-9b98-42ec-acb6-028e20403cfb.png)

Reply:
OK now I know where I read that (when we discussed about it at Ricardo and caching node_modules was giving me bad vibes) ^^

I guess we should listen to: 
![image](https://user-images.githubusercontent.com/2678610/143936761-da34f620-6f92-4430-9b7e-fbc5d4ecbab1.png)

Reply:
This is only the case because cypress has a binary that it download outside node_modules. We can cache both as far as i understand.

Reply:
Yes we can try this

Reply:
When you have some time you can have a look. I think this solution is aligned with their recommendations 

https://glebbahmutov.com/blog/do-not-let-cypress-cache-snowball/